File: Python/fixedPoint.py
import numpy as np
from math import *

# Global variables
NMAX = 20
ERROR = 100
TOLERANCE = 0.00001


def fixedPoint(p0, tol, nmax, g, test):
    iteri = 0
    # Ejecucion
    while iteri < nmax:
        p = g(p0)
        if not test:
            print("p = \t" + str(p))
        error = abs(p0 - p)
        if error < tol:
            break
        p0 = p
        if not test:
            print("p0 = \t" + str(p0))
            print()
        iteri += 1

    # Salida
    if iteri > nmax:
        print("El metodo no converge")
    else:
        if not test:
            print("El punto de la funcion g(x) es \t" + str(p))
        else:
            return p


def g_test_1(x):
    return (2 - (np.exp(x)) + x**2) / 3


if __name__ == "__main__":
    p0 = 0
    fixedPoint(p0, TOLERANCE, NMAX, g_test_1, False)


# Tests
def test_1():
    p0 = 0
    assert fixedPoint(p0, TOLERANCE, NMAX, g_test_1, True) == 0.25753180626753985


"""
# Entradas
ecuacion = input("Ingrese la ecuacion \n")
p0 = float(input("Ingrese el punto inicial \n"))
tol = float(input("ingrese la tolerecia \n"))
nmax = float(input("Ingrese el numero maximo de iteracciones \n"))
iteri = 0


def g(x):
    return eval(ecuacion)

"""

File: C++/fixedPoint.cpp
#include <math.h>
#include <stdio.h>

#include <cstdlib>
// Global variables
#define e 2.718281828
#define NMAX 100
#define TOLERANCE 0.00001
#define EXACTITUDEC 0.00005

double fun(double);

int main() {
    double P0, TOL, P, b;
    int i, N, OK;

    P0 = 0;
    TOL = 0.00001;
    N = 10;

    // step 1
    i = 1;
    OK = 1;
    // step 2
    printf("\n\nI\t\tP\n");
    while (i <= N && OK == 1) {
        // step 3
        P = fun(P0);
        printf("%d\t%12.8f\n", i, P);
        // step 4
        b = fabs(P - P0);
        if (b < TOL) {
            printf("Approximate solution P = %12.8f\n", P);
            printf("Number of iterations N = %d\n", i);
            printf("Tolerance TOL = %5.11f\n\n", TOL);
            OK = 0;
        }

        else {
            // step 5
            i++;
            // step 6
            P0 = P;
        }
    }
    // step 7
    if (OK == 1)
        printf("\n\nEl metodo fallo con un numero de iteraciones N = %d\n\n", N);

    return 0;
}

double fun(double x) {
    return (2 - pow(e, x) + x * x) / 3;
}